### ISSUE TITLE
Clarify use case of app & recall/precision

### ISSUE THREAD

**Issue by mueslo** [nextcloud/suspicious_login#461]
Hi, 
I have had this app installed for a while now, but cannot really figure out what it is supposed to be used for, for multiple reasons.

as far as I can tell, the app

i) collects (ip, uid) tuples of sucessful logins and trains a ML model from that
ii) checks if the login is classified as suspicious
iii) if yes, sends a notification to the account in question

However, really, what is the point of (iii)?
It does not block logins. And in fact, once an attacker gets into the account, they can just delete that notification. Which will then also disappear on all other devices. The only useful thing about it would be if 'suspicious'-classified logins could not delete any notifications from suspicious_login or could not login at all without 2FA or something.

2nd, about point ii). In my case it seems to flag basically EVERY* log-in with a new IP as suspicious, even if it comes from an IP address very close (numerically) to previous IP addresses. If I want it to do that, there is no need for any sort machine learning. It seems the model may not be 100% fit for purpose. I assume it also uses neither geographical information nor internet topology information, which would be very useful for a classification model pipeline (i.e. turn (ip, uid) into (ip, uid, lat, long) for training with geographic info), and would drastically reduce the amount of data required. Is there a command with which I can test the classifier on an (ip, uid) tuple of my choosing?

3rd, again about point ii). The security page lists two stats, first, recall ('fraction of all suspicious logins captured', TP/(TP+FN), precision ('fraction of logins classified as suspicious that are indeed suspicious', TP/(TP+FP). How can you make these stats? I would wager that 100% of the successful logins on my instance are legitmate so there are zero "actually suspicious" logins, so the real precision is zero. According to the repository description you are also not using failed logins. So where do you get authoritative/"supervised" information from? All you really know is (TP+FP) and (TN+FN).
So, what exactly do TP, FP, FN mean here? My best guess:

Assume G is the full address space, M: (ip,uid) -> {0, 1} is your classifier model.
My best guess is that you 1) assume all captured (successful) logins L are legitimate, and that 2) all** G \ L are suspicious, of which you pick a random set. You then run M(t) on the non-training part of set L (and get FP & TN), and run M(t) on a random sample of G \ L (and get TP & FN).

Again, is there a command with which I can test just the trained classifier on an (ip, uid) tuple of my choosing?

*"So far the app has captured 555049 logins (including client connections), of which 829 are distinct (IP, UID) tuples."

**or a certain subset with minimum distance to L according to some metric?

**Comment by mueslo** [comment]
After trying this add-on for multiple years, I have to be a bit blunt and say it just straight up **doesn't seem to work as intended**. Since there are not really any configuration options, I don't think it is a configuration error, but must be an error in the model or programming. Maybe it's due to the small number of users on my instance, but no idea.

Basically *every* new login from a new IP address yields a warning, even if it is in the same /24 subnet as previous attempts. But more than that, *the exact IPs* I have logged in from in the past are regularly classified as suspicious on new logins! (At least once every 6 months, sometimes once a month) This strongly implies something is fundamentally broken.

![image](https://user-images.githubusercontent.com/847751/185116351-959427c5-a15b-4392-97f1-2c77a1401884.png)

On average I get two "suspicious login" e-mails per week (see image). None of those are actually suspicious (precision = 0). When the failure rate is so high, it just becomes another warning you ignore. Roughly one quarter of the warnings are logins via my mobile carrier. All logins via my mobile carrier come from the same narrow /22 subnet.

![image](https://user-images.githubusercontent.com/847751/185128255-bf13810f-3f3e-4715-99f2-3fe48bf40859.png)

**Comment by mueslo** [comment]
My Nextcloud instance's trained model on average claims roughly 94% precision and 90% recall. 

Precision = TP / (TP + FP)
Recall = TP / (TP + FN)

Based on your previous explanation I assume you just calculate it with the random sampling of all IPs and assume that all random IPs that trigger the model are TP, random IPs that don't trigger the model are FN, captured login IPs that trigger the model are FP and captured login IPs that don't trigger the model are TN.

TP: random trigger
FN: random notrigger
FP: hist trigger
TN: hist notrigger

In this model, 94% is not a good precision at all, since that means If I just keep logging in from random IPs I have logged in from in the past, I have a roughly 6% chance to get a warning, in other words roughly every 17th login.

This further reinforces my thinking that the chosen model is just not appropriate. A simple hardcoded IP address distance check would probably work leaps and bounds better than this.

And further:

>> I assume it also uses neither geographical information nor internet topology information, which would be very useful for a classification model pipeline (i.e. turn (ip, uid) into (ip, uid, lat, long) for training with geographic info)

> IP addresses have topology if you look at them as vector of bits. The neural net is able to learn those. For ipv6 we also throw away the lower half of the vector.

Their topology is not good enough to not produce a lot of false positives. IPs that are close in number tend to be close geographically, that much is true. But the inverse is not true, IPs that are close geographically can be very far away from each other, due to different IP blocks per carrier/company/institution, and even sometimes multiple distant IP blocks within a single institution.

Also, the training approach you mentioned is bad in general and needs to overfit to appear to work properly. It's like training an image recognition net to detect cats, and as negatives giving random images - including cats. The resulting neural net will not detect cats, it will detect specific images of cats.
